### ISSUE TITLE
Add application filter support for vnc environment

### ISSUE THREAD

**Issue by hitswint** [k0kubun/xremap#69]
Xremap works under vnc environment, but cannot filter different software. I noticed x-set-keys (https://github.com/kawao/x-set-keys), which supports filter for window class under vnc. So there must be a way for Xremap to support similar thing. Thanks so much for this great software and your hard work!

**Comment by k0kubun** [comment]
Could you show your config and the entire output of `RUST_LOG=debug xremap` from the point where the command starts to the point where the problem is reproduced while switching applications? Also, have you checked if the same thing works if you don't use VNC?

Since it just seems to use the same kinds of APIs at least for detecting the current application, as long as you use the X11 binary of xremap and configure `application` appropriately, it should work.

**Comment by hitswint** [comment]
I used the emacs.yml in the example folder and modified it as:
```
  - name: Emacs
    application:
      not: Emacs
    remap:
       .....
```
The output includes the list of all devices and "application-client: none (supported: false)".
I installed the binary file using:
cargo install xremap
I tried "cargo install xremap --features x11", but failed with following errors:

```
error[E0308]: mismatched types
  --> /home/swint/.cargo/registry/src/github.com-1ecc6299db9ec823/xremap-0.2.3/src/client/x11_client.rs:74:64
   |
74 |                         wm_class = std::ffi::CString::from_raw(x_class_hint.res_class as *mut i8)
   |                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u8`, found `i8`
   |
   = note: expected raw pointer `*mut u8`
              found raw pointer `*mut i8`

For more information about this error, try `rustc --explain E0308`.
error: failed to compile `xremap v0.2.3`, intermediate artifacts can be found at `/tmp/cargo-installucSZtH`

Caused by:
  could not compile `xremap` due to previous error
```

**Comment by hitswint** [comment]
My platform is arm64, which might be the reason for the compiling error.

**Comment by k0kubun** [comment]
> I installed the binary file using:
> cargo install xremap

This is the cause. See also: https://github.com/k0kubun/xremap#dynamic-binding

I don't have any local arm64 environment or one on CI, so I cannot investigate that myself. What is the version of your rustc? I use rustc 1.57.0, and using the same version might help. If not, I'll check if we can distribute arm64 binary from CI.

**Comment by hitswint** [comment]
So does the xremap without X11 feature support application filter?

**Comment by k0kubun** [comment]
> So does the xremap without X11 feature support application filter?

It doesn't need to be X11, but you should have at least one feature to use `application` filters. Since `x-set-keys` is working in the environment, you should enable the X11 feature.

> I have followed the options in the dynamic-binding section, but no luck.

Those options are about starting the `xremap` command after successfully installing it. So don't worry about them for now.

---

Could you also try this?

```
git clone https://github.com/k0kubun/xremap
cd xremap
cargo build --features x11
```

And use `target/build/xremap`. If `cargo build --features x11` fails, I'd like to see the full output of the command as well.

**Comment by hitswint** [comment]
"cargo build --release --features x11" outputs:
'''
error[E0308]: mismatched types
  --> src/client/x11_client.rs:74:64
   |
74 |                         wm_class = std::ffi::CString::from_raw(x_class_hint.res_class as *mut i8)
   |                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u8`, found `i8`
   |
   = note: expected raw pointer `*mut u8`
              found raw pointer `*mut i8`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `xremap` due to previous error
'''
I know almost nothing about rust, what do you mean by "use target/release/xremap"? I can find the target/release directory, but no xremap file.

**Comment by k0kubun** [comment]
> I can find the target/release directory, but no xremap file.

What you guessed is correct. It doesn't exist because you encountered a build failure.

Meanwhile, could you try doing the same thing on the current `arm64` branch https://github.com/k0kubun/xremap/tree/arm64? It looks like the issue is Arm-specific, but cross-compilation doesn't work because of the X11 library, and GitHub Actions doesn't seem to support Arm64. 

For actually supporting it in the future, I'm thinking about changing the X11 library to something more portable, but now I'm trying to fix the build failure on your environment through the error message reported here. So please keep posting the entire output of `cargo build --release --features x11` to prepare a branch that works for Arm64.

**Comment by hitswint** [comment]
Everything works fine this time. No compilation errors and xremap can filter different application as expected. Thanks for your hard work!!!
One more thing, i have used xkeysnail on other platform, and it supports a feature called "escape_next_key". The feature allows user to input the original keybinding with a prefix C-q. It's really useful when i want to use the default keybinding of some application. I did not see such things in the config file of xremap. So does the xremap support that?

**Comment by k0kubun** [comment]
It has not been implemented since I wasn't sure if there's anybody who wants it. If there's a legitimate use case for it, I can add that.

> It's really useful when i want to use the default keybinding of some application.

Why don't you just define a keymap for that application though? If you do, you don't need to press `C-q` in the first place. Are there too many keymaps to be reverse-overridden? Could you show an example config and a situation with a particular application where escape_next_key is useful?

**Comment by hitswint** [comment]
For example, I use a word processing software called wps, in which C-i and C-b were originally mapped to italic and bold font. But in emacs keybindings, those keys were also used for navigation. So I have to figure out some other keybindings which are not conflict with existing ones. Some applications come with a lot of default keybindings, it's really bored to do such things for each app. So it will be better if there is a way to trigger its default keybindings.

**Comment by k0kubun** [comment]
Ah, do you mean you use both "C-i"/"C-b" and "C-q C-i"/"C-q C-b" in wps? Then it makes sense to me.

**Comment by hitswint** [comment]
Yes, this only happens when the default keybindings of some application conflicts with emacs keymap.

**Comment by k0kubun** [comment]
Could you check if this branch works as you expect https://github.com/k0kubun/xremap/pull/74?

FYI, I deleted the `arm64` branch since I committed https://github.com/k0kubun/xremap/commit/ee93bb41f931658da5d578961b7e3b954a0e8ff0 to master. If you cannot build the master branch or the pull request's branch, please let me know.

**Comment by k0kubun** [comment]
Thank you for confirming it. I released the feature as v0.2.5. I'm thinking about supporting arm64 binary distribution in the future since I learned about https://github.com/cross-rs/cross, but for now, please continue to just build yours from the master branch.

**Comment by k0kubun** [comment]
Hi, I supported arm64 (aarch64) binary releases at v0.5.0. You might want to try that when you have a chance.
